### PR TITLE
Avoid having contended mutexes on same cacheline

### PR DESCRIPTION
**Pull request by redbaron**
CPUs have to serialise write access to a single cache line
effectively reducing level of possible parallelism. Placing
mutexes on different cache lines avoids this problem.

Most gains will be seen on NUMA servers where CPU interconnect
traffic is especially expensive

Before:
```
go test . -run none -bench BenchmarkFingerprintLocker
BenchmarkFingerprintLockerParallel-4   	 2000000	       932 ns/op
BenchmarkFingerprintLockerSerial-4     	30000000	        49.6 ns/op
```

After:
```
go test . -run none -bench BenchmarkFingerprintLocker
BenchmarkFingerprintLockerParallel-4   	 3000000	       569 ns/op
BenchmarkFingerprintLockerSerial-4     	30000000	        51.0 ns/op
```

### CODE REVIEW

**Comment by juliusv** [comment]
Ha, I've been wanting to do this for a while, thanks :)

**Comment by redbaron** [comment]
I  don't really get why there are 1024 mutexes minimum and 4096 mutexes by default, but maybe these numbers can be lowered now?

**Comment by redbaron** [comment]
or possibly embed mutexes into metric (or wrapped metric struct) to be more cache friendly. From the look of it every time mutex is accessed in local/storage.go there is a matching access to a metric itself, so keeping then adjastent is natural then.

You'd also wont need hash + modulo anymore as every metric would have its own lock, making locking more fain grained

**Comment by juliusv** [comment]
Embedding the locks in the metrics wouldn't work, as there can be multiple `model.Metric`s for the same time series flying around, and we also need to be able to lock a series according to its metric fingerprint before the series exists. Also, then we'd have millions of locks instead of just a couple - not a huge deal, but more memory wasted.

No idea about the 1024 / 4096 minimum / maximum. @beorn7 can probably say more about that.

Introducing the padding is great though.

**Comment by beorn7** [comment]
> No idea about the 1024 / 4096 minimum / maximum.

The 1024 is just a safety guard so that nobody accidentally sets a lower number and gets sever performance penalties. The rationale is that a lower number would save so little RAM that it doesn't make sense to go down there.

The 4096 default is there because it is still pretty cheap in terms of RAM but it helps a lot for high frequency ingestion with a decent number of time series. (With more time series, you need to tweak this value much higher.)

With this change, the RAM consumption increases by 8x. Still I think we should not go below 4096 by default as this is what "normal" users will need anyway.

File: storage/local/locker.go [hunk 1]
```diff
@@ -19,6 +19,12 @@ import (
 	"github.com/prometheus/common/model"
 )
 
+// avoid false sharing when using array of mutexes
```
Comment (redbaron) [inline storage/local/locker.go]: fixed

File: storage/local/locker.go [hunk 2]
```diff
@@ -19,6 +19,12 @@ import (
...
+type paddedMutex struct {
+	sync.Mutex
+	pad [64]byte
```
Comment (redbaron) [inline storage/local/locker.go]: fixed

**Comment by juliusv** [comment]
:+1: thanks!

**Comment by redbaron** [comment]
@beorn7 my concern wasn't RAM usage, but L1 cache size which is just 32KB

**Comment by beorn7** [comment]
Feel free to play with lower numbers. My rationale so far is that you need high performance if you have high frequency injection and/or high number of series, resulting in a lot of concurrency. Reducing the number of FP mutexes will create more contention, which will probably kill the caching effects of using fewer locks. I could be wrong, of course.

**Pull request merged**